build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -0,0 +1,129 @@
+import org.gradle.api.plugins.JavaLibraryPlugin;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.compile.CompileOptions;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.gradle.jvm.tasks.Jar;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.JavaToolchainService;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.tree.ClassNode;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.inject.Inject;
+
+import static org.objectweb.asm.Opcodes.V_PREVIEW;
+
+public class MrjarPlugin implements Plugin<Project> {
+
+    private static final Pattern JAVA_SOURCESET_PATTERN = Pattern.compile("java(\\d{2})");
+
+    private final JavaToolchainService javaToolchains;
+
+    @Inject
+    MrjarPlugin(JavaToolchainService javaToolchains) {
+        this.javaToolchains = javaToolchains;
+    }
+
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().apply(JavaLibraryPlugin.class);
+        var javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+
+        var srcDir = project.getProjectDir().toPath().resolve("src");
+        try (var subdirStream = Files.list(srcDir)) {
+            for (Path sourceset : subdirStream.toList()) {
+                String sourcesetName = sourceset.getFileName().toString();
+                Matcher sourcesetMatcher = JAVA_SOURCESET_PATTERN.matcher(sourcesetName);
+                if (sourcesetMatcher.matches()) {
+                    int javaVersion = Integer.parseInt(sourcesetMatcher.group(1));
+                    addMrjarSourceset(project, javaExtension, sourcesetName, javaVersion);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void addMrjarSourceset(Project project, JavaPluginExtension javaExtension, String sourcesetName, int javaVersion) {
+        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
+        GradleUtils.extendSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME, sourcesetName);
+
+        Jar jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME).get();
+        jarTask.dependsOn(sourceSet.getCompileJavaTaskName());
+        jarTask.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+        jarTask.manifest(manifest -> {
+            manifest.attributes(Map.of("Multi-Release", "true"));
+        });
+
+        JavaCompile compileTask = project.getTasks().withType(JavaCompile.class).named(sourceSet.getCompileJavaTaskName()).get();
```

Comment:
we should use project.getTasks().withType(JavaCompile.class).named(sourceSet.getCompileJavaTaskName()).configure here to avoid realisation of the task.